docs/proposals/remote-artifacts.md

```diff
@@ -0,0 +1,267 @@
+*   Manage remote artifacts;
+*   Amend container images created by the given `BuildStrategy`;
+*   Walking a remote directory tree, similarly to a `git checkout`;
+
+# Proposal
+
+The enhancement proposal is centered around the idea of declaring external dependencies, here
+called "artifacts", and being able to use them during the container image building process.
+
+## Build Artifacts
+
+The remote artifacts will be directly expressed in the body of a `Build` resource, as the
+following example:
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: Build
+metadata:
+  name: nexus-jar
+spec:
+  artifacts:
+    - name: main-jar
+      url: https://nexus.company.com/app/main.jar
+      path: $(workspace)/classpath/main.jar
+```
+
+A new attribute, .spec.artifacts will be added, containing a slice of types with:
+
+*   **Name**: The actual name of the artifact, optional;
+*   **URL**: Universal location of the artifact;
+*   **Path**: Represents the final artifact location with placeholder (`$(workspace)`) support;
+
+The resource `UID`/`GID` will be defined by the same user running the Tekton task, later on, we
+can extend the API to support arbitrary configuration.
+
+### Standalone CRD
+
+Alternatively, we may define the artifacts as a standalone CRD, that is a `BuildArtifact`
+resource. The advantage of this design is being able to exchange, and reuse, artifacts on several
+builds. For instance, if two projects are sharing a common logo image, both `Builds` will refer to
+the same `BuildArtifact`.
+
+The following snippet shows how an Artifact (`BuildArtifact`) will be represented. In this
+example, consider a Java based application with Jar hosted in Nexus, that needs to be added to
+application's `classpath` directory.
+
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: BuildArtifact
+metadata:
+  name: nexus-jars
+spec:
+  artifacts:
+    - name: main-jar
+      url: https://nexus.company.com/app/main.jar
+      path: $(workspace)/classpath/main.jar
+```
+
+## Using Artifacts
+
+Giving the artifacts declared, a developer will be able to include those artifacts in Build
+resources as per the following example:
+
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: Build
+metadata:
+  name: java-application
+spec:
+  artifacts:
+    - nexus-jars
+```
+
+By adding spec.artifacts, the operator will generate build steps to download external data, and
+place it in the expected location before the build strategy.Furthermore, developers will be able
+to overwrite artifacts on `BuildRun` level, allowing for more use-cases on which different sets of
+artifacts is required.
+
+
+## Steps and Helper
+
+Artifacts are requirements for the build process, thus downloading and preparing artifacts must
+happen before the build process starts. To achieve the objective, the operator will generate a new
+task step to download the artifacts.
+
+The download may happen using existing open-source software, more specifically `wget`. The
```

Comment:
The "download" looks critical. We get this for free from tekton, Wouldn´t it make more sense to do this natively with "golang" or to see if Tekton have already something for us? e.g. they do not use a "git" binary for pulling, they have their own git [pkg](https://github.com/tektoncd/pipeline/blob/master/pkg/git/git.go)

Reply:
I understand what Tekton does for the VCS (source-code), but what would it do for remote artifacts? Could we use Tekton to download them as well?

Reply:
It does not appear such a resource type exists. If we accept this proposal, I recommend submitting a TEP to see if we can get upstream Tekton buy-in.

Reply:
I was mainly referring to do a native implementation in go, I would like to avoid introducing extra binaries dependencies in the code. We might not need Tekton at all.

Reply:
I would like to explore the Tekton resource idea further, for the record.

Reply:
I think we all agree this is not enforcing the usage of `wget` and that another more native implementation(e.g. tekton way) might be reused. Can we mention this somewhere?

Reply:
Can we leverage the Tekton source to enable storage or something first?:
https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#resource-types

I think `wget` is a little weak. How abour the private resource (with auth) or ftp?

Reply:
`PipelineResources` at this time are most likely not going to get out of alpha stage.  There has been a fair amount of consternation and churn around how exactly how to move forward, though.

But minimally I would not advise a `PipelineResource` styled path.

That said, if I had to pick, I'd say "custom tasks" might be the most likely path along the "tekton resource" angle.  Also, submitting something "similar" but different most likely would get the response of "look at custom tasks".

See https://github.com/tektoncd/community/blob/master/teps/0002-custom-tasks.md for details.

Reply:
there is also a "specializing task" which has some literature around it but now TEP as of yet

Reply:
Yeah, a specific resource in upstream Tekton would be a hard sell at the moment given the direction `PipelineResources` is taking.